### PR TITLE
Pages

### PR DESCRIPTION
**Pull request by andreabedini**
This is a first go at generating html pages from the input specification. At this time I am generating a simple html page for each package version included in the index.

TODO:
- [ ] list of packages
- [ ] list of version per packages
- [ ] change log (i.e. list of revisions)

### CODE REVIEW

**Review by michaelpj — COMMENTED** [review]
Just some drive-by things

File: app/Distribution/Aeson.hs [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Distribution.Aeson where
```
[the remark this answers is not in the document]
Comment (andreabedini) [reply app/Distribution/Aeson.hs:12]: Working on it! https://github.com/andreabedini/Cabal-syntax-json

File: app/Distribution/PackageDescription/FieldGrammar/Copy.hs [hunk 2]
```diff
@@ -0,0 +1,159 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+--
+-- This module is a partial copy of `Distribution.PackageDescription.FieldGrammar`.
+--
+-- It is needed because that module does not export CompatFilePath and
+-- CompatLicenseFile, which are needed to call packageDescriptionFieldGrammar
```
[the remark this answers is not in the document]
Comment (andreabedini) [reply app/Distribution/PackageDescription/FieldGrammar/Copy.hs:10]: > upstream issue?

definitely

> Can you use cabal-syntax eventually?

I am, this module is ripped off `Cabal-syntax-json` see previous comment.

File: cabal.project [hunk 3]
```diff
@@ -1,2 +1,3 @@
 packages: .
 index-state: 2022-08-29T00:00:00Z
+with-compiler: ghc-8.10.7
```
[the remark this answers is not in the document]
Comment (andreabedini) [reply cabal.project:3]: As much as I can agree, I have moved by default ghc to 9.2 and `tomland` does not support it yet.

File: app/Distribution/PackageDescription/FieldGrammar/Copy.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreabedini) [reply app/Distribution/PackageDescription/FieldGrammar/Copy.hs:10]: Fixed upstream https://github.com/haskell/cabal/pull/8523

**Comment by andreabedini** [comment]
@michaelpj @angerman Any feedback?

Test with `nix run github:andreabedini/foliage/pages build` from the cardano-haskell-packages repo. Website is in `_repo` along with the rest.

**Comment by michaelpj** [comment]
I cleared out `_cache` and it worked, but I didn't even realise there was a cache before now!

**Comment by michaelpj** [comment]
Currently the index table has package-versions as its entries rather than packages. I can see why, this way you can search the whole thing etc. But I think it'll get pretty unwieldy as we get more stuff (especially since the `source` cells take so much space!). 

Maybe we could also support Hackage's structure of having a page for the package that then lists all the published versions is nice. (Or rather, it links you to the page for the latest version and then the page for each version has links to the pages for each version). We could add that later, though.

The links for package-versions on the index page go to e.g. file:///home/michael/projects/iohk/cardano-haskell-packages/_repo/package/ouroboros-consensus-cardano-tools-0.1.0.0 rather than file:///home/michael/projects/iohk/cardano-haskell-packages/_repo/package/ouroboros-consensus-cardano-tools-0.1.0.0/index.html, so you get a directory listing instead of the actual page.

"Home" took me to the root directory of my filesystem! Maybe this stuff works properly when you're not opening it from the filesystem? :thinking: 

What can go in "Type" other than "package"? A tooltip would not go amiss (applies everywhere tbh!).

"Timestamp" of what? How about "Time added" or something?

Otherwise LGTM!

**Comment by andreabedini** [comment]
> I cleared out `_cache` and it worked, but I didn't even realise there was a cache before now!

:-D yeah, I need to be more careful with the caching.

Thanks for the detailed feedback, I'll go through it today.

@phadej I am keen to hear what you think about this.

**Comment by andreabedini** [comment]
> Currently the index table has package-versions as its entries rather than packages. 

I might rename that page to "timeline", so we can have a page with just the last versions.

> Maybe we could also support Hackage's structure of having a page for the package that then lists all the published versions is nice. 

Yes we could do that. I would only need to pass the list of other versions to the page version page.

> The links for package-versions on the index page go to e.g. file:///home/michael/projects/iohk/cardano-haskell-packages/_repo/package/ouroboros-consensus-cardano-tools-0.1.0.0

This is because you have been using file:// I thought browsers didn't even support that anymore, does the search box work in that case. I can fix some of this but not all.

> What can go in "Type" other than "package"? A tooltip would not go amiss (applies everywhere tbh!).

"Revision". Good idea about the tool tip.

> "Timestamp" of what? How about "Time added" or something?

Yes, time added. Thanks!

**Comment by michaelpj** [comment]
> I might rename that page to "timeline", so we can have a page with just the last versions.

"Repository's entries" (I just noticed that the title is inconsistent and says "Repository's content") is actually fairly descriptive, it just didn't quite convey it to me. Perhaps just "All package versions"? And then "All packages" for just the packages?

> This is because you have been using file:// I thought browsers didn't even support that anymore, does the search box work in that case.

The whole thing works fine, including the search box. It's just that some of the links don't work properly! FWIW my expectation of most static site-producing things is that they work when opened from the filesystem. That's true of sites built with mdbook or sphinx, for example. It's nice for testing, since you don't need to run a webserver.

**Comment by michaelpj** [comment]
Oh yes, also is there a plan to incorporate a user-provided landing page? We have that at the moment and it's quite nice. With this branch it's gone. Haddock has `--prologue` that lets you put some text at the start of the landing page, we could copy that. Since probably you do want to fit it into the overall site structure that you're building here.

**Comment by andreabedini** [comment]
> Oh yes, also is there a plan to incorporate a user-provided landing page?

I was going to let the user deal with it, like `mv index.html summary.html` not the best but it should work for the moment.

**Pull request merged**